src/react/hooks/use-presence.ts

```diff
@@ -0,0 +1,160 @@
+import { useChatConnection } from './use-chat-connection.js';
+import { useRoom } from './use-room.js';
+
+/**
+ * The options for the {@link usePresence} hook.
+ */
+export interface UsePresenceParams extends StatusParams {
+  /**
+   * The data to enter the room with. Any JSON serializable data can be provided.
+   */
+  data?: PresenceData;
+}
+
+export interface UsePresenceResponse extends ChatStatusResponse {
+  /**
+   * A shortcut to the {@link Presence.update} method.
+   */
+  readonly update: Presence['update'];
+
+  /**
+   * A shortcut to the {@link Presence.leave} method.
+   */
+  readonly leave: Presence['leave'];
+
+  /**
+   * Indicates whether the current user is present in the room.
+   */
+  readonly isPresent: boolean;
+
+  /**
+   * Indicates if an error occurred while entering or leaving the room.
+   */
+  readonly error?: ErrorInfo;
+
+  /**
+   * Provides access to the underlying {@link Presence} instance of the room.
+   */
+  readonly presence: Presence;
+}
+
+const INACTIVE_CONNECTION_STATES = new Set<ConnectionLifecycle>([
+  ConnectionLifecycle.Suspended,
+  ConnectionLifecycle.Failed,
+]);
+
+/**
+ * A hook that provides access to the {@link Presence} instance in the room.
+ * It will use the instance belonging to the room in the nearest {@link ChatRoomProvider} in the component tree.
+ * On calling, the hook will enter the room with the provided data and leave the room when the component unmounts.
+ * The {@link isPresent} flag will indicate when the user has become present in the room.
+ *
+ * @param params - Allows the registering of optional callbacks.
+ * @returns UsePresenceResponse - An object containing the {@link Presence} instance and methods to interact with it.
+ */
+export const usePresence = (params?: UsePresenceParams): UsePresenceResponse => {
+  const { currentStatus: connectionStatus, error: connectionError } = useChatConnection({
+    onStatusChange: params?.onConnectionStatusChange,
+  });
+  const { room, roomError, roomStatus } = useRoom({
+    onStatusChange: params?.onRoomStatusChange,
+  });
+
+  const [isPresent, setIsPresent] = useState(false);
+  const [error, setError] = useState<ErrorInfo | undefined>();
+
+  // store the roomStatus in a ref to ensure the correct value is used in the effect cleanup
+  const roomStatusAndConnectionStatusRef = useRef({ roomStatus, connectionStatus });
+
+  // we can't use the data param directly in a dependency array as it will cause an infinite loop
+  const dataRef = useRef(params?.data);
+  useEffect(() => {
+    dataRef.current = params?.data;
+  }, [params?.data]);
+
+  useEffect(() => {
+    // Update the ref when roomStatus changes
+    roomStatusAndConnectionStatusRef.current = { roomStatus, connectionStatus };
+  }, [roomStatus, connectionStatus]);
+
+  // enter the room when the hook is mounted
+  useEffect(() => {
+    const canJoinPresence = roomStatus === RoomLifecycle.Attached && !INACTIVE_CONNECTION_STATES.has(connectionStatus);
```

Comment:
We need a check here to see if a user has explicitly left presence - the `useEffect` will run every time the room status changes, but if they've explicitly left presence, we shouldn't automatically re-enter them.

Either that, or we remove the `leave` method and handle the lifecycle of being present by the lifecycle of the hook (i.e. you can update your presence data whenever you like, but if you don't want to remain present, you should stop calling the hook).

Reply:
I think the idea of removing it here all together keeps us more in line with how ably-js do things, so I will go with that approach. Also, it feel a bit more intuitive that way too.